### PR TITLE
CDK-871: Expose CSVFileReader in the SPI.

### PR DESCRIPTION
**Pull request by rdblue**
This makes the CSVFileReader public and adds an InputFormat constructor,
rather than just a FileSystem and Path constructor. Because the
InputFormat may not be a FSDataInputStream, this also adds a check when
attempting to use the reader as a Hadoop RecordReader.

### CODE REVIEW

**Comment by joey** [comment]
The patch looks fine, but I think this is another example of where we need a properly defined SPI that includes these kind of classes that aren't meant for general use, but are meant for external consumers.

**Comment by rdblue** [comment]
> The patch looks fine, but I think this is another example of where we need a properly defined SPI that includes these kind of classes that aren't meant for general use, but are meant for external consumers.

I agree.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java [hunk 1]
```diff
@@ -87,11 +101,17 @@ public void initialize() {
...
-    } catch (IOException ex) {
-      throw new DatasetReaderException("Cannot open path: " + path, ex);
+    if (incoming != null) {
+      this.size = 1; // avoid divide by 0 errors
+    } else {
+      Preconditions.checkNotNull(fs, "FileSystem cannot be null");
+      Preconditions.checkNotNull(path, "Path cannot be null");
+      try {
+        this.incoming = fs.open(path);
+        this.size = fs.getFileStatus(path).getLen();
```
Comment (tomwhite) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java]: What if the file length is zero? Wouldn't you still get a divide-by-zero error later?

BTW this is a bug fix that is independent of the issue being fixed here so should be fixed separately - in a different commit, if not a separate JIRA.

**Comment by tomwhite** [comment]
The SPI and the internal implementation classes have been conflated and both are in the spi package. We could move everything that isn't being used externally to an impl package (we already have a precedent for those), then over time add things to spi only when it has been designed with external use in mind. Thoughts?

**Comment by joey** [comment]
I think that's definitely the direction we want to head towards. There's still a conflation of SPI versus limited audience APIs. In my mind, the SPI has the interfaces that are needed to extend Kite with a new format or to make use of a new storage API. Limited audience APIs are APIs that aren't meant for end-users, but are meant for other frameworks that need to integrate Kite, like Flume or Sqoop.

Should we explore an annotation based approach like Hadoop has adopted?

**Comment by tomwhite** [comment]
I have reservations about the Hadoop annotations, since they are widely ignored. The problem is that you can't see the annotation at the point of use (e.g. in an IDE autocomplete situation). If you have a class in an `impl` package then users are less likely to import them into their code.

What's an example of something that is a limited audience API that can't go in the public API or the SPI?

I opened https://issues.cloudera.org/browse/CDK-878 to discuss this further.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java (hunk 1 above)
Comment (rdblue) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java]: I opened [CDK-882](https://issues.cloudera.org/browse/CDK-882) for this.

**Comment by rdblue** [comment]
With CDK-882 and CDK-878 filed, I don't think there are any remaining issues before this can be merged.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java (hunk 1 above)
Comment (tomwhite) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVFileReader.java]: There's still a comment ("avoid divide by 0 errors") which suggests that there are no divide-by-zero errors.

**Pull request merged**